File: backend-node/src/services/authService.ts
import { sha256, hashPassword, comparePassword } from "../utils/hash.js";
import { randomBytes } from "crypto";
import jwt from "jsonwebtoken";
import { env } from "../config/env.js";
import { issueDigitalID } from "./BlockchainService.js";
import { randomUUID } from "crypto";
import { processLocation } from "./AnomalyService.js";
import {
  getAllTourists,
  getTouristById,
  getTouristByEmail,
  createTourist,
  updateTourist,
  type ITourist,
} from "./mongoStore.js";
import { TouristModel } from "../schemas/index.js";

type TouristRegistration = {
  name: string;
  email: string;
  phone: string;
  passportNumber: string;
  dateOfBirth?: string;
  address?: string;
  gender?: string;
  nationality?: string;
  emergencyContact?: { name?: string; phone?: string };
  bloodType?: string;
  allergies?: string[];
  medicalConditions?: string[];
  passwordHash: string; // Actually raw password from frontend
  currentLat?: number;
  currentLng?: number;
};

export async function registerTourist(
  payload: TouristRegistration
): Promise<{ ok: true; tourist: ITourist } | { ok: false; message: string }> {
  const existing = await getTouristByEmail(payload.email);
  if (existing) {
    return { ok: false, message: "Email already registered" };
  }

  const rawPassword = payload.passwordHash;
  const hashedPassword = await hashPassword(rawPassword);
  const idHashInput = `${payload.passportNumber}${payload.phone}${new Date().toISOString()}`;
  const idHash = sha256(idHashInput);

  const now = new Date();
  const expiry = new Date(now.getTime());
  expiry.setFullYear(expiry.getFullYear() + 1);

  const tourist = await createTourist({
    _id: randomUUID(),
    name: payload.name,
    email: payload.email,
    phone: payload.phone,
    passportNumber: payload.passportNumber,
    dateOfBirth: payload.dateOfBirth,
    address: payload.address,
    gender: payload.gender,
    nationality: payload.nationality,
    emergencyContact: payload.emergencyContact,
    bloodType: payload.bloodType,
    allergies: payload.allergies,
    medicalConditions: payload.medicalConditions,
    passwordHash: hashedPassword,
    idHash,
    idExpiry: expiry.toISOString(),
    lastSeen: now.toISOString(),
    safetyScore: 100,
    currentLat: payload.currentLat,
    currentLng: payload.currentLng,
  });

  await issueDigitalID(tourist._id, tourist.idHash!);
  return { ok: true, tourist };
}

export async function validateTouristLoginByEmail(email: string, rawPassword: string) {
  const tourist = await getTouristByEmail(email);
  if (!tourist) {
    return null;
  }
  const isValid = await comparePassword(rawPassword, tourist.passwordHash);
  if (!isValid) {
    return null;
  }
  return tourist;
}

export function login(touristId: string) {
  return jwt.sign({ sub: touristId, role: "tourist" }, env.jwtSecret, { expiresIn: env.jwtExpiry as any });
}

export async function getProfile(touristId: string) {
  return getTouristById(touristId);
}

export async function updateProfile(touristId: string, payload: Record<string, unknown>) {
  const tourist = await getTouristById(touristId);
  if (!tourist) {
    return null;
  }

  const email = payload.email as string | undefined;
  if (email) {
    const existingWithEmail = await getTouristByEmail(email);
    if (existingWithEmail && existingWithEmail._id !== touristId) {
      throw new Error("Email already registered");
    }
  }

  const updates: Partial<ITourist> = {};
  if (payload.name !== undefined) updates.name = payload.name as string;
  if (payload.email !== undefined) updates.email = payload.email as string;
  if (payload.phone !== undefined) updates.phone = payload.phone as string;
  if (payload.passportNumber !== undefined) updates.passportNumber = payload.passportNumber as string;
  if (payload.dateOfBirth !== undefined) updates.dateOfBirth = payload.dateOfBirth as string;
  if (payload.address !== undefined) updates.address = payload.address as string;
  if (payload.gender !== undefined) updates.gender = payload.gender as string;
  if (payload.nationality !== undefined) updates.nationality = payload.nationality as string;
  if (payload.emergencyContact !== undefined) {
    updates.emergencyContact = payload.emergencyContact as { name?: string; phone?: string };
  }
  if (payload.bloodType !== undefined) updates.bloodType = payload.bloodType as string;
  if (payload.allergies !== undefined) updates.allergies = payload.allergies as string[];
  if (payload.medicalConditions !== undefined) {
    updates.medicalConditions = payload.medicalConditions as string[];
  }

  return updateTourist(touristId, updates);
}

export async function verifyIdHash(idHash: string) {
  const tourist = await TouristModel.findOne({ idHash }).lean();
  if (!tourist) {
    throw new Error("Digital ID not found or invalid.");
  }
  return tourist;
}

export async function updateLocation(touristId: string, lat: number, lng: number, accuracy?: number, speed?: number, heading?: number) {
  const tourist = await getTouristById(touristId);
  if (!tourist) {
    throw new Error("Tourist not found.");
  }

  const updated = await updateTourist(touristId, {
    currentLat: lat,
    currentLng: lng,
    speed,
    heading,
    locationAccuracy: accuracy,
    lastSeen: new Date().toISOString(),
  });

  if (updated) {
    // Write to location log (time-series) for ML pipeline
    const { createLocationLog } = await import("./mongoStore.js");
    await createLocationLog({
      touristId,
      latitude: lat,
      longitude: lng,
      speed,
      heading,
      accuracy,
      safetyScoreAtTime: updated.safetyScore ?? 100,
    }).catch((err) => console.error("Location log write failed:", err));

    await processLocation(updated, accuracy);
  }

  return updated;
}

export async function listTourists() {
  return getAllTourists();
}

export async function requestPasswordReset(email: string) {
  const tourist = await getTouristByEmail(email);
  if (!tourist) {
    return { ok: true };
  }

  const resetToken = randomBytes(20).toString("hex");
  const resetTokenHash = sha256(resetToken);
  const expires = new Date(Date.now() + 1000 * 60 * 30);

  await updateTourist(tourist._id, {
    resetTokenHash,
    resetTokenExpires: expires,
  });

  return { ok: true, resetToken };
}

export async function confirmPasswordReset(token: string, newPassword: string) {
  const tokenHash = sha256(token);
  const tourist = await TouristModel.findOne({
    resetTokenHash: tokenHash,
    resetTokenExpires: { $gt: new Date() },
  });

  if (!tourist) {
    return { ok: false, message: "Invalid or expired reset token." };
  }

  const hashedPassword = await hashPassword(newPassword);
  tourist.passwordHash = hashedPassword;
  tourist.resetTokenHash = undefined;
  tourist.resetTokenExpires = undefined;
  await tourist.save();

  return { ok: true };
}

File: backend-spring/src/main/java/com/safarsathi/service/AuthService.java
package com.safarsathi.service;

import com.safarsathi.dto.TouristRegistrationRequest;
import com.safarsathi.dto.TouristResponse;
import com.safarsathi.entity.Tourist;
import com.safarsathi.repository.TouristRepository;
import com.safarsathi.security.JwtService;
import com.safarsathi.util.HashUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final TouristRepository touristRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final BlockchainService blockchainService;
    private final AnomalyService anomalyService;

    /**
     * Register a new tourist.
     */
    public Map<String, Object> registerTourist(TouristRegistrationRequest request) {
        if (touristRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new IllegalArgumentException("Email already registered");
        }

        String hashedPassword = passwordEncoder.encode(request.getPasswordHash());
        String idHashInput = request.getPassportNumber() + request.getPhone() + Instant.now().toString();
        String idHash = HashUtil.sha256(idHashInput);
        Instant expiry = Instant.now().plus(365, ChronoUnit.DAYS);

        Tourist.EmergencyContact ec = null;
        if (request.getEmergencyContact() != null) {
            ec = new Tourist.EmergencyContact(
                    request.getEmergencyContact().get("name"),
                    request.getEmergencyContact().get("phone"));
        }

        Tourist tourist = Tourist.builder()
                .id(UUID.randomUUID().toString())
                .name(request.getName())
                .email(request.getEmail())
                .phone(request.getPhone())
                .passportNumber(request.getPassportNumber())
                .dateOfBirth(request.getDateOfBirth())
                .address(request.getAddress())
                .gender(request.getGender())
                .nationality(request.getNationality())
                .emergencyContact(ec)
                .bloodType(request.getBloodType())
                .allergies(request.getAllergies())
                .medicalConditions(request.getMedicalConditions())
                .passwordHash(hashedPassword)
                .idHash(idHash)
                .idExpiry(expiry.toString())
                .currentLat(request.getCurrentLat())
                .currentLng(request.getCurrentLng())
                .lastSeen(Instant.now().toString())
                .safetyScore(100.0)
                .build();

        tourist = touristRepository.save(tourist);

        // Issue digital ID on mock blockchain
        blockchainService.issueDigitalID(tourist.getId(), idHash);

        String token = jwtService.generateToken(tourist.getId(), "tourist");

        Map<String, Object> result = new HashMap<>();
        result.put("touristId", tourist.getId());
        result.put("qr_content", "/api/admin/id/verify?hash=" + idHash);
        result.put("token", token);
        result.put("user", toResponse(tourist));
        return result;
    }

    /**
     * Login a tourist with email and password.
     */
    public Map<String, Object> loginTourist(String email, String password) {
        Tourist tourist = touristRepository.findByEmail(email)
                .orElse(null);
        if (tourist == null || !passwordEncoder.matches(password, tourist.getPasswordHash())) {
            return null;
        }

        String token = jwtService.generateToken(tourist.getId(), "tourist");

        Map<String, Object> result = new HashMap<>();
        result.put("touristId", tourist.getId());
        result.put("qr_content", "/api/admin/id/verify?hash=" + tourist.getIdHash());
        result.put("token", token);
        result.put("user", toResponse(tourist));
        return result;
    }

    /**
     * Get tourist profile.
     */
    public TouristResponse getProfile(String touristId) {
        return touristRepository.findById(touristId)
                .map(this::toResponse)
                .orElse(null);
    }

    /**
     * Update tourist profile.
     */
    public TouristResponse updateProfile(String touristId, Map<String, Object> payload) {
        Tourist tourist = touristRepository.findById(touristId).orElse(null);
        if (tourist == null) return null;

        if (payload.containsKey("email")) {
            String newEmail = (String) payload.get("email");
            touristRepository.findByEmail(newEmail).ifPresent(existing -> {
                if (!existing.getId().equals(touristId)) {
                    throw new IllegalArgumentException("Email already registered");
                }
            });
            tourist.setEmail(newEmail);
        }
        if (payload.containsKey("name")) tourist.setName((String) payload.get("name"));
        if (payload.containsKey("phone")) tourist.setPhone((String) payload.get("phone"));
        if (payload.containsKey("passportNumber")) tourist.setPassportNumber((String) payload.get("passportNumber"));
        if (payload.containsKey("dateOfBirth")) tourist.setDateOfBirth((String) payload.get("dateOfBirth"));
        if (payload.containsKey("address")) tourist.setAddress((String) payload.get("address"));
        if (payload.containsKey("gender")) tourist.setGender((String) payload.get("gender"));
        if (payload.containsKey("nationality")) tourist.setNationality((String) payload.get("nationality"));
        if (payload.containsKey("bloodType")) tourist.setBloodType((String) payload.get("bloodType"));

        if (payload.containsKey("emergencyContact")) {
            @SuppressWarnings("unchecked")
            Map<String, String> ecMap = (Map<String, String>) payload.get("emergencyContact");
            if (ecMap != null) {
                tourist.setEmergencyContact(new Tourist.EmergencyContact(ecMap.get("name"), ecMap.get("phone")));
            }
        }
        if (payload.containsKey("allergies")) {
            @SuppressWarnings("unchecked")
            List<String> allergies = (List<String>) payload.get("allergies");
            tourist.setAllergies(allergies);
        }
        if (payload.containsKey("medicalConditions")) {
            @SuppressWarnings("unchecked")
            List<String> mc = (List<String>) payload.get("medicalConditions");
            tourist.setMedicalConditions(mc);
        }

        tourist = touristRepository.save(tourist);
        return toResponse(tourist);
    }

    /**
     * Update tourist location and trigger anomaly detection.
     */
    public void updateLocation(String touristId, Double lat, Double lng, Double accuracy) {
        Tourist tourist = touristRepository.findById(touristId)
                .orElseThrow(() -> new RuntimeException("Tourist not found."));
        tourist.setCurrentLat(lat);
        tourist.setCurrentLng(lng);
        tourist.setLastSeen(Instant.now().toString());
        tourist = touristRepository.save(tourist);
        anomalyService.processLocation(tourist);
    }

    /**
     * Verify an ID hash (for QR code scanning by admin).
     */
    public Tourist verifyIdHash(String idHash) {
        return touristRepository.findByIdHash(idHash)
                .orElseThrow(() -> new RuntimeException("Digital ID not found or invalid."));
    }

    /**
     * List all tourists.
     */
    public List<Tourist> listTourists() {
        return touristRepository.findAll();
    }

    /**
     * Request a password reset token.
     */
    public Map<String, Object> requestPasswordReset(String email) {
        Map<String, Object> result = new HashMap<>();
        result.put("acknowledged", true);

        Tourist tourist = touristRepository.findByEmail(email).orElse(null);
        if (tourist == null) return result;

        String resetToken = UUID.randomUUID().toString().replace("-", "").substring(0, 40);
        String resetTokenHash = HashUtil.sha256(resetToken);
        Instant expires = Instant.now().plus(30, ChronoUnit.MINUTES);

        tourist.setResetTokenHash(resetTokenHash);
        tourist.setResetTokenExpires(expires);
        touristRepository.save(tourist);

        result.put("resetToken", resetToken);
        return result;
    }

    /**
     * Confirm password reset with token.
     */
    public boolean confirmPasswordReset(String token, String newPassword) {
        String tokenHash = HashUtil.sha256(token);
        Tourist tourist = touristRepository
                .findByResetTokenHashAndResetTokenExpiresAfter(tokenHash, Instant.now())
                .orElse(null);
        if (tourist == null) return false;

        tourist.setPasswordHash(passwordEncoder.encode(newPassword));
        tourist.setResetTokenHash(null);
        tourist.setResetTokenExpires(null);
        touristRepository.save(tourist);
        return true;
    }

    /**
     * Convert Tourist entity to TouristResponse DTO.
     */
    public TouristResponse toResponse(Tourist tourist) {
        Map<String, String> ecMap = null;
        if (tourist.getEmergencyContact() != null) {
            ecMap = new HashMap<>();
            ecMap.put("name", tourist.getEmergencyContact().getName());
            ecMap.put("phone", tourist.getEmergencyContact().getPhone());
        }

        return TouristResponse.builder()
                .id(tourist.getId())
                .name(tourist.getName())
                .email(tourist.getEmail())
                .phone(tourist.getPhone())
                .passportNumber(tourist.getPassportNumber())
                .dateOfBirth(tourist.getDateOfBirth())
                .address(tourist.getAddress())
                .gender(tourist.getGender())
                .nationality(tourist.getNationality())
                .emergencyContact(ecMap)
                .bloodType(tourist.getBloodType())
                .allergies(tourist.getAllergies())
                .medicalConditions(tourist.getMedicalConditions())
                .currentLat(tourist.getCurrentLat())
                .currentLng(tourist.getCurrentLng())
                .lastSeen(tourist.getLastSeen())
                .idHash(tourist.getIdHash())
                .idExpiry(tourist.getIdExpiry())
                .build();
    }
}
